### DISCUSSION TITLE
Reference setup with Prisma, NextJS, and local Postgres / PGBouncer?

### DISCUSSION THREAD

**Discussion by a-type** [prisma/prisma#11987] [Q&A]
I've got NextJS setup with Prisma, which of course means there will be the standard Serverless issues to account for.

In a deployed environment we're using Supabase and their built-in PGBouncer based connection pooling. Haven't seen anything go wrong there.

But locally, I'm constantly crashing with connection pool issues. I have a local database and I'm running PGBouncer in front of it.

Has someone gotten this working stably? Can you share some details of how you've structured your `DATABASE_URL` or any extra logic you've added to Next to cleanup Prisma? Here's my DATABASE_URL:

`DATABASE_URL="postgresql://local:local@localhost:6543/local?pgbouncer=true&connection_limit=1"`

I still get these errors pop up constantly:

```
[app]   Timed out fetching a new connection from the connection pool. (More info: http://pris.ly/d/connection-pool, Current connection limit: 1)
[app]     at cb (/......../node_modules/@prisma/client/runtime/index.js:36958:17)
[app]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[app]     at async getSessionAndUser (/......../node_modules/@next-auth/prisma-adapter/dist/index.js:22:36) {
[app]   name: 'GetSessionAndUserError',
[app]   code: undefined
[app] }
```

**Comment by austincrim** [comment]
Hey @a-type 👋 ,

Have you implemented [this logic](https://www.prisma.io/docs/support/help-articles/nextjs-prisma-client-dev-practices#problem) for dealing with live reload issues when using Prisma?

```ts
import { PrismaClient } from '@prisma/client'

declare global {
  // allow global `var` declarations
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined
}

export const prisma =
  global.prisma ||
  new PrismaClient({
    log: ['query'],
  })

if (process.env.NODE_ENV !== 'production') global.prisma = prisma
```

**Reply by a-type** [reply]
Nope, that's probably it. Thanks for the help, I'll try it tomorrow.

**Reply by a-type** [reply]
I will say, with my limited understanding of the module system in Node, it's confusing to me how this would help. Assigning the value to `global` would make it accessible between modules, but since it's only being referenced within this module's scope, I'm not sure how it will behave differently from a simple `export const prisma = new PrismaClient()`. But serverless function contexts are pretty confusing anyway. Is the idea that the `global` context might be preserved between separate serverless contexts invoking this same module, whereas if it were only in module scope it might be instantiated multiple times?

**Comment by janpio** [comment]
Also: Why `connection_limit=1`? That limits your application to only execute 1 query at a time.

**Reply by a-type** [reply]
It was the recommended setting for serverless function environments. I assume because each function has its own context, so even sharing Prisma within that context might exhaust connections if multiple different functions are running at once and each is allowed 9 connections? Increasing the limit did not avoid the issue, or even seem to reduce its frequency.

**Reply by janpio** [reply]
A limit of 1 enforces that multiple queries are _always_ running after each other. Hence it is more probably that any query will hit the timeout and you will get the error message you posted above.

When you are using PgBouncer, you do not care that much about the # of open connections from Prisma, as Prisma only talks to PgBouncer which then actually manages a pool of open connections to the database. Do we need to make that clearer in the docs?

**Reply by a-type** [reply]
I see. So is that limit causing Prisma to queue its own connections to PGBouncer, which in turn queues those connections to Postgres?

I returned to the docs to review the serverless recommendations, and your advice is definitely reflected there. I didn't return to that section after setting up PGBouncer, so I missed that detail. I should probably have read over that whole page again with my current setup in mind.

I think perhaps it got confusing with the advice for both non-pooled and pooled serverless environments being interwoven. Some sections make it clear which context they refer to, but sometimes there are comments [like this](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#:~:text=Note%3A%20Setting%20the%20connection_limit%20to%201%20in%20serverless%20environments%20is%20a%20recommended%20starting%20point%2C%20but%20this%20value%20can%20also%20be%20tuned.) which don't specify the context of the recommendation (pooled or not), which shake my confidence as I'm still absorbing the information and trying to reason about how it fits my setup. Given how different the constraints are between pooled/non-pooled, I wonder if separating them entirely is appropriate.

I think in this case I reached that point where everything just starts to feel like 'magic incantations,' I can't see any demonstrable effects of my changes, and I live in fear of the next crash popping up at any time... when I get into that mode I start to second guess if my `connection_limit` really should be 1, etc.

That's why I reached out for a concrete example specific to my setup. It helps to have a specific point of reference when the choices start to feel very arbitrary and opaque.

**Reply by janpio** [reply]
That makes sense. We should probably go over these docs again and see where we need to qualify any advice, or put it into a bigger context.

**Comment by a-type** [comment]
Neither suggestions have relieved the problem, but a bit more logging indicates it's probably my fault (and likely to affect our deployed app as well) - our app is bursting a large number of requests to authenticate a session, which really shouldn't be happening. It would make sense that this is overwhelming the pooling, even with limits raised and PGBouncer in place.

I'm still really thankful for the suggestions, as now I can feel confident that the problem is with our code rather than mistakes I made in configuration. Thanks!

**Reply by colinchadwick** [reply]
Hey there! Did you find any solution? I’m running into the same issue, using Next JS with Next Auth and Prisma. I also have the feeling that it has something to do with the authentication. Thanks!

**Reply by colinchadwick** [reply]
Implementing this logic solves the issue: 

```
import { PrismaClient } from '@prisma/client'

const prisma = global.prisma || new PrismaClient();

if (process.env.NODE_ENV === "development") global.prisma = prisma;

export default prisma
```

After that just import the Prisma client from that file.

**Reply by a-type** [reply]
Our solution to excessive queries was to migrate to JWT-based sessions instead of storing sessions in the database.

**Comment by Venipa** [comment]
im doing this in nextjs
```ts
import { PrismaClient } from "@prisma/client";

let client: PrismaClient;
export default function getPrismaInstance() {
  return client ?? (client = new PrismaClient());
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
